spark/src/main/java/org/apache/iceberg/spark/source/SparkAppenderFactory.java

```diff
@@ -86,4 +98,81 @@
       throw new RuntimeIOException(e);
     }
   }
+
+  @Override
+  public DataWriter<InternalRow> newDataWriter(EncryptedOutputFile file, FileFormat fileFormat, StructLike partition) {
+    return new DataWriter<>(
+        newAppender(file.encryptingOutputFile(), fileFormat), fileFormat,
+        file.encryptingOutputFile().location(), spec, partition, file.keyMetadata());
+  }
+
+  @Override
+  public EqualityDeleteWriter<InternalRow> newEqDeleteWriter(EncryptedOutputFile outputFile, FileFormat format,
+                                                             StructLike partition) {
+    try {
+      switch (format) {
+        case PARQUET:
+          return Parquet.writeDeletes(outputFile.encryptingOutputFile())
+              .createWriterFunc(msgType -> SparkParquetWriters.buildWriter(dsSchema, msgType))
+              .overwrite()
+              .rowSchema(writeSchema)
+              .withSpec(spec)
+              .withPartition(partition)
+              .equalityFieldIds(equalityFieldIds)
+              .withKeyMetadata(outputFile.keyMetadata())
+              .buildEqualityWriter();
+
+        case AVRO:
+          return Avro.writeDeletes(outputFile.encryptingOutputFile())
+              .createWriterFunc(ignored -> new SparkAvroWriter(dsSchema))
+              .overwrite()
+              .rowSchema(writeSchema)
+              .withSpec(spec)
+              .withPartition(partition)
+              .equalityFieldIds(equalityFieldIds)
+              .withKeyMetadata(outputFile.keyMetadata())
+              .buildEqualityWriter();
+
+        default:
+          throw new UnsupportedOperationException("Cannot write unsupported format: " + format);
+      }
+
+    } catch (IOException e) {
+      throw new UncheckedIOException("Failed to create new equality delete writer", e);
+    }
+  }
+
+  @Override
+  public PositionDeleteWriter<InternalRow> newPosDeleteWriter(EncryptedOutputFile outputFile, FileFormat format,
+                                                              StructLike partition) {
+    try {
+      switch (format) {
+        case PARQUET:
+          return Parquet.writeDeletes(outputFile.encryptingOutputFile())
+              .createWriterFunc(msgType -> SparkParquetWriters.buildWriter(dsSchema, msgType))
```

Comment:
Here we will need to provide the schema which combines with `writeSchema`  and file & pos columns to construct the `ParquetValueWriter` because the `PositionDeleteStructWriter` will treat the  user-provided rows and file+pos as a whole record to write into the target file.

Reply:
The builder here will automatically wrap the row schema and the function passed here to add the extra schema layer. So we just need to configure this for rows, not for the combined schema. That's part of why the position writer's delete method accepts file and position independent of row, to keep the encapsulation and not leak the concern to places like this.

Reply:
In this sentence `reateWriterFunc(msgType -> SparkParquetWriters.buildWriter(dsSchema, msgType))`,  the `msgType`  have included the `file` and `pos` columns, but the `dsSchema`  hasn't. I mean we need to provide a  correct `dsSchema` so that it could match the `msgType`  exactly.  It's similar to this: https://github.com/apache/iceberg/pull/1663/files#diff-7f498f01885f6e813bc3892c8dfb02b8893365540438b78b3a0221f9c8667c8fR211

Reply:
I think that's a bug. My intent was to have the caller configure the writer just like normal for a data file or equality delete file. There is no need to expose that complexity to the writer. So we should update the builder to extract the row type and pass it into the `createWriterFunc`.